From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gerson Fernando Budke <gerson.budke@ossystems.com.br>
Subject: [PATCH] [GH#414] STM32_ADC_Common: Implement EOCAL interrupt

The STM32L0 ADC still in a loop waiting the calibration phase
finish. This implement the calibration interrupt to finish
calibration procedure and allow ADC driver to work.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

Ported to Renode v1.17.0 STM32_ADC_Common API.

diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs
@@ @@ public STM32_ADC_Common(IMachine machine, double referenceVoltage, uint external
             ADCChannelCount = channelCount;
             WatchdogCount = watchdogCount;
             this.hasChannelSelect = hasChannelSelect;
+            this.hasCalibration = hasCalibration;
             this.hasChannelInjection = hasChannelInjection;
             this.resolutionRange = resolutionRange;
             this.hasChannelPreselection = hasChannelPreselection;
@@ @@ private void UpdateInterrupts()
                 irq |= endOfConversionInjectedFlag.Value && endOfConversionInjectedInterruptEnable.Value;
                 irq |= endOfSequenceInjectedFlag.Value && endOfSequenceInjectedInterruptEnable.Value;
             }
+            if(hasCalibration)
+            {
+                irq |= endOfCalibrationFlag.Value && endOfCalibrationInterruptEnable.Value;
+            }
             IRQ.Set(irq);
         }
 
@@ @@ private Dictionary<long, DoubleWordRegister> BuildRegistersMap(bool hasCalibrati
             if(hasCalibration)
             {
                 isrRegister
-                    .WithTaggedFlag("EOCAL", 11)
+                    .WithFlag(11, out endOfCalibrationFlag, FieldMode.Read | FieldMode.WriteOneToClear, name: "EOCAL")
                     // Simplified logic - hardware delays LDORDY until voltage regulator settles.
                     .WithFlag(12, valueProviderCallback: _ => adcRegulatorEnable.Value, name: "LDORDY");
                 interruptEnableRegister
-                    .WithTaggedFlag("EOCALIE", 11)
+                    .WithFlag(11, out endOfCalibrationInterruptEnable, name: "EOCALIE")
                     .WithTaggedFlag("LDORDYIE", 12);
             }
             else
@@ @@ private Dictionary<long, DoubleWordRegister> BuildRegistersMap(bool hasCalibrati
                         }, name: "ADSTP")
                     .WithFlag(28, out adcRegulatorEnable, name: "ADVREGEN")
                     .WithReservedBits(29, 2)
-                    .WithTaggedFlag("ADCAL", 31);
+                    .WithFlag(31, valueProviderCallback: _ => false, writeCallback: (_, __) =>
+                        {
+                            UpdateInterrupts();
+                        }, name: "ADCAL");
 
             if(hasLinearityCalibration)
             {
@@ @@ private ushort ResolutionToBits(Resolution resolution)
         private IValueRegisterField data;
         private IFlagRegisterField analogWatchdogSingleChannel;
         private IFlagRegisterField endOfSequenceInterruptEnable;
+        private IFlagRegisterField endOfCalibrationInterruptEnable;
         private IFlagRegisterField endOfSamplingInterruptEnable;
         private IFlagRegisterField endOfConversionInterruptEnable;
         private IFlagRegisterField[] analogWatchdogsInterruptEnable;
         private IFlagRegisterField adcReadyInterruptEnable;
         private IFlagRegisterField adcOverrunInterruptEnable;
         private IFlagRegisterField endOfSequenceFlag;
+        private IFlagRegisterField endOfCalibrationFlag;
         private IFlagRegisterField endOfConversionFlag;
         private IFlagRegisterField[] analogWatchdogFlags;
         private IFlagRegisterField adcReadyFlag;
@@ @@ private ushort ResolutionToBits(Resolution resolution)
         private readonly IDMA dma;
         private readonly int dmaChannel;
         private readonly bool hasChannelSelect;
+        private readonly bool hasCalibration;
         private readonly ResolutionRange resolutionRange;
         private readonly bool hasChannelPreselection;
         private readonly uint externalEventFrequency;
